### ISSUE TITLE
Support for FidelityFX Super Resolution upscaler

### ISSUE THREAD

**Issue by gudvinr** [alvr-org/ALVR#780]
Since AMD released [FSR source code](https://gpuopen.com/fidelityfx-superresolution/) it is now possible to use it.

It is specifically designed to be used for real-time rendering in games, which makes FSR a viable option for [upscaling alternative](https://github.com/alvr-org/ALVR/wiki/Real-time-video-upscaling-experiments).

Although currently there is no public information on whether it is compatible with XR1 or not, FSR is vendor-independent and Vulkan is supported.

This might not produce good quality in every case, however. For example, AMD mentions that image must be antialiased and noise-free. Given that source frames are video frames from lossy codec, this should be definitely tested and probably can't be used as the only upscaling option.

For the best quality it is preferable to upscale _before_ encoding but it will inevitable increase bandwidth requirement.

**Comment by NightHammer1000** [comment]
I am not sure if this can be done at ALVRs Level.
Because it needs to be done between Geometry/PostFX Rendering and UI Rendering. So basicly in Engine, Not beyond that for Optimal Results. For a SteamVR Universal Solution that does Upscale on top the Finished Render (which is far from best practice) see: https://github.com/fholger/openvr_fsr

**Comment by gudvinr** [comment]
> Because it needs to be done between Geometry/PostFX Rendering and UI Rendering.

Yes, it is why I mentioned that "this might not produce good quality in every case". However, VR is quite good case for post processing done outside since it is meant to be immersive and many games either do not have traditional HUD or use HUD elements tied to gloves, and thus rendered as floating surfaces rather than OSD.

> see: https://github.com/fholger/openvr_fsr

They mention that it requires manual assist on per-game basis and it won't work for games that do not use D3D11 and openvr. I believe that openxr is goto standard for future games since it is backed by basically every big player in this field, so it kinda important thing to consider too.

**Comment by NightHammer1000** [comment]
Still. It wont change anything if FSR is Implemented into ALVR because ALVR only works with the Already Finished Image.
Its the same with fixed foveated rendering in ALVR.

FSR needs to be done at Engine Level.
So its in the Hands of the Devs to use it.

**Comment by zarik5** [comment]
OpenVR allows games to pass a depth map to be used by the drivers, but 1) I don't know if a depth map can be plugged in FSR, 2) ALVR would need to upscale the image before encoding or send both lower res color frames and depth maps, diminishing the performance benefit.

**Comment by gudvinr** [comment]
> FSR needs to be done at Engine Level.

IIRC there's no hard requirement for that. It does not use additional information (such as motion vectors in DLSS) and is basically just fancy compute shader that only needs one frame at a time.

FSR is recommended to be implemented inside engine only because any post-processing effects that introduce noise (e.g. film grain) can mess up edge detection pass in that shader. So if there's no heavy post-processing in the game, it should not produce otherwise trash results
